libs/libcommon/src/libcommon/simple_cache.py

```diff
@@ -127,6 +128,19 @@ class CacheEntryDoesNotExistError(DoesNotExist):
     pass
 
 
+def increase_metrics(kind: str, http_status: HTTPStatus, error_code: Optional[str] = None) -> None:
```

Comment:
Since we now only change the metrics incrementally, both datasets (cache and metrics) must be aligned. Maybe we should run the complete query (snapshot) regularly to ensure we are not diverging. In particular, at startup, maybe the cache dataset is already filled while the metrics one is empty.

Not sure if it's needed, but the idea is the same as how PostgreSQL writes on the disk (https://www.postgresql.org/docs/current/wal-configuration.html): it appends logs (Write-Ahead Logging - WAL) incrementally, but also does regular Checkpoints to ensure integrity.

Reply:
Is it worth storing the metric history? or should we keep only the last record in the metric collection as it is currently implemented?
I think Prometheus will store already the history so maybe only option two is needed ( Correct -> Increment/decrement the current last record  that is the last captured snapshot value)

Reply:
> Or, even, we could run the cron job more often, ie. every 5 minutes, but:: the cron job would only update one metric at a time, to reduce the query complexity. For each metric, we would store the date of the last "checkpoint", and the cron job would update the first "outdated" metric (ie. the last date was more than 1 day ago)

In this case, we could be having the same problem as we currently have: high ratio scanned keys vs returned keys, even if we reduce the group of metrics to calculate. The query complexity would reduce but anyway, we will end with a query targeting issue (e.g for split-first-rows-from-streaming we have like 158287 records, and with the aggregated record we retrieve like 10 results with a ratio of 15828 - exceeding the threashold that is 1000).


I think the first option is better and enough for now:

> So, for the cron job, we could keep the current main code, but reduce the frequency from 2 minutes to 1 day, for example (we don't need to fix more often than that, I think -- it's only to avoid the metrics to diverge too much). By the way, it would be nice to have a report, when the cron job finishes, of how much it has corrected the values.

Reply:
Hmmm, I think the opposite:

-> option 1: we do a big aggregate query once a day: it will raise the same warning (too many scanned keys)
-> option 2: once a day at most (for each cache type) we run a query with a filter on the specific cache type. It should use the index and not raise the warning (I think). The trick here is that we would get the cache types from the entries in the metrics collections, so: no need to query the cache collection to get the list of cache types.

wdyt?

Reply:
> -> option 2: once a day at most (for each cache type) we run a query with a filter on the specific cache type. It should use the index and not raise the warning (I think)

Currently, the problematic query already uses the appropriate index. I think there is not a performance problem with the execution plan (a bad plan or something like that), but it is a problem with the way/design we process metrics (querying everything to calculate aggregated data) and I think that is what the query ratio wants to highlight.
If we do option 2, we could end up with a lower query targeting value but anyway, it will always be more than 1000 so we could still receive the alert every time a cache kind is processed (maybe 24 times a day in the worst cases 1 by each processing step).
In my opinion, it could be better to have option 1 since anyway we will have the alert but in this case, it will be just once a day in the worst case.

BTW, I think the traditional data processing systems face this problem in a batch incremental way over periods of time, let's say saving the last_read_timestamp and computing the metrics every x time just for the records updated after last_read_timestamp. Maybe we could change to something like that in the future?

Reply:
OK, let's do option 1, and create a different issue to try to optimize it later.